Crash NPE: When editor activity is recreated after being killed in background

Happens with both Visual and Aztec editors.

Steps to reproduce:

1. Turn "Don't keep activities" on.
2. Open any draft or published post (new post does not cause this crash).
3. While at the editor screen press Home button to minimize the app.
4. Return to the app, notice crash.

```
FATAL EXCEPTION: main
   Process: org.wordpress.android, PID: 20465
   java.lang.RuntimeException: Unable to start activity ComponentInfo{org.wordpress.android/org.wordpress.android.ui.posts.EditPostActivity}: java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.String org.wordpress.android.fluxc.model.PostModel.getPostFormat()' on a null object reference
       at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2646)
       at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2707)
       at android.app.ActivityThread.-wrap12(ActivityThread.java)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1460)
       at android.os.Handler.dispatchMessage(Handler.java:102)
       at android.os.Looper.loop(Looper.java:154)
       at android.app.ActivityThread.main(ActivityThread.java:6077)
       at java.lang.reflect.Method.invoke(Native Method)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:865)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:755)
    Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.String org.wordpress.android.fluxc.model.PostModel.getPostFormat()' on a null object reference
       at org.wordpress.android.ui.posts.EditPostSettingsFragment.fetchSiteSettingsAndUpdateDefaultPostFormatIfNecessary(EditPostSettingsFragment.java:168)
       at org.wordpress.android.ui.posts.EditPostSettingsFragment.onActivityCreated(EditPostSettingsFragment.java:154)
       at android.app.Fragment.performActivityCreated(Fragment.java:2361)
       at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1014)
       at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1171)
       at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1153)
       at android.app.FragmentManagerImpl.dispatchActivityCreated(FragmentManager.java:2039)
       at android.app.FragmentController.dispatchActivityCreated(FragmentController.java:176)
       at android.app.Activity.performCreateCommon(Activity.java:6658)
       at android.app.Activity.performCreate(Activity.java:6666)
       at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1118)
       at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2599)
```

--- a/WordPress/build.gradle
+++ b/WordPress/build.gradle
@@ -134,7 +134,7 @@ dependencies {
     apt 'com.google.dagger:dagger-compiler:2.0.2'
     provided 'org.glassfish:javax.annotation:10.0-b28'
 
-    compile ('com.github.wordpress-mobile.WordPress-FluxC-Android:fluxc:feature~add-get-post-by-remote-id-SNAPSHOT') {
+    compile ('com.github.wordpress-mobile.WordPress-FluxC-Android:fluxc:6603e2cf5c58f764169e842c676b14fb7ff01fdb') {
         exclude group: "com.android.volley";
     }
 

--- a/WordPress/build.gradle
+++ b/WordPress/build.gradle
@@ -134,7 +134,7 @@ dependencies {
     apt 'com.google.dagger:dagger-compiler:2.0.2'
     provided 'org.glassfish:javax.annotation:10.0-b28'
 
-    compile ('com.github.wordpress-mobile.WordPress-FluxC-Android:fluxc:6422c5020f442bbe42003cb1ae887b8f898e1d2d') {
+    compile ('com.github.wordpress-mobile.WordPress-FluxC-Android:fluxc:feature~add-get-post-by-remote-id-SNAPSHOT') {
         exclude group: "com.android.volley";
     }
 

--- a/WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java
+++ b/WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java
@@ -320,7 +320,10 @@ public class EditPostActivity extends AppCompatActivity implements
                 mPost.setStatus(PostStatus.PUBLISHED.toString());
             } else if (extras != null) {
                 // Load post passed in extras
-                initializePostObjectsWithLocalId(extras.getInt(EXTRA_POST_LOCAL_ID));
+                mPost = mPostStore.getPostByLocalPostId(extras.getInt(EXTRA_POST_LOCAL_ID));
+                if (mPost != null) {
+                    initializePostObject();
+                }
             }
         } else {
             mDroppedMediaUris = savedInstanceState.getParcelable(STATE_KEY_DROPPED_MEDIA_URIS);
@@ -329,9 +332,11 @@ public class EditPostActivity extends AppCompatActivity implements
             // if we have a remote id saved, let's first try with that, as the local Id might have changed
             // after FETCH_POSTS
             if (savedInstanceState.containsKey(STATE_KEY_POST_REMOTE_ID)) {
-                initializePostObjectsWithRemoteId(savedInstanceState.getLong(STATE_KEY_POST_REMOTE_ID));
+                mPost = mPostStore.getPostByRemotePostId(savedInstanceState.getLong(STATE_KEY_POST_REMOTE_ID), mSite);
+                initializePostObject();
             } else if (savedInstanceState.containsKey(STATE_KEY_POST_LOCAL_ID)) {
-                initializePostObjectsWithLocalId(savedInstanceState.getInt(STATE_KEY_POST_LOCAL_ID));
+                mPost = mPostStore.getPostByLocalPostId(savedInstanceState.getInt(STATE_KEY_POST_LOCAL_ID));
+                initializePostObject();
             }
 
             mEditorFragment = (EditorFragmentAbstract) fragmentManager.getFragment(savedInstanceState, STATE_KEY_EDITOR_FRAGMENT);
@@ -409,27 +414,15 @@ public class EditPostActivity extends AppCompatActivity implements
         ActivityId.trackLastActivity(ActivityId.POST_EDITOR);
     }
 
-    private void initializePostObjectsWithLocalId(int localPostId) {
-        mPost = mPostStore.getPostByLocalPostId(localPostId);
-        if (mPost != null) {
-            initializePostObject();
-        }
-    }
-
-    private void initializePostObjectsWithRemoteId(long remotePostId) {
-        mPost = mPostStore.getPostByRemotePostId(remotePostId, mSite);
-        if (mPost != null) {
-            initializePostObject();
-        }
-    }
-
     private void initializePostObject() {
-        mOriginalPost = mPost.clone();
-        mPost = UploadService.updatePostWithCurrentlyCompletedUploads(mPost);
-        mMediaMarkedUploadingOnStartIds =
-                AztecEditorFragment.getMediaMarkedUploadingInPostContent(this, mPost.getContent());
-        Collections.sort(mMediaMarkedUploadingOnStartIds);
-        mIsPage = mPost.isPage();
+        if (mPost != null) {
+            mOriginalPost = mPost.clone();
+            mPost = UploadService.updatePostWithCurrentlyCompletedUploads(mPost);
+            mMediaMarkedUploadingOnStartIds =
+                    AztecEditorFragment.getMediaMarkedUploadingInPostContent(this, mPost.getContent());
+            Collections.sort(mMediaMarkedUploadingOnStartIds);
+            mIsPage = mPost.isPage();
+        }
     }
 
     // this method aims at recovering the current state of media items if they're inconsistent within the PostModel.

--- a/WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java
+++ b/WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java
@@ -549,7 +549,9 @@ public class EditPostActivity extends AppCompatActivity implements
         // Saves both post objects so we can restore them in onCreate()
         savePostAsync(null);
         outState.putInt(STATE_KEY_POST_LOCAL_ID, mPost.getId());
-        outState.putLong(STATE_KEY_POST_REMOTE_ID, mPost.getRemotePostId());
+        if (!mPost.isLocalDraft()) {
+            outState.putLong(STATE_KEY_POST_REMOTE_ID, mPost.getRemotePostId());
+        }
         outState.putBoolean(STATE_KEY_IS_NEW_POST, mIsNewPost);
         outState.putSerializable(WordPress.SITE, mSite);
 

--- a/WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java
+++ b/WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java
@@ -330,8 +330,7 @@ public class EditPostActivity extends AppCompatActivity implements
             // after FETCH_POSTS
             if (savedInstanceState.containsKey(STATE_KEY_POST_REMOTE_ID)) {
                 initializePostObjectsWithRemoteId(savedInstanceState.getLong(STATE_KEY_POST_REMOTE_ID));
-            }
-            else if (savedInstanceState.containsKey(STATE_KEY_POST_LOCAL_ID)) {
+            } else if (savedInstanceState.containsKey(STATE_KEY_POST_LOCAL_ID)) {
                 initializePostObjectsWithLocalId(savedInstanceState.getInt(STATE_KEY_POST_LOCAL_ID));
             }
 

--- a/WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java
+++ b/WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java
@@ -171,6 +171,7 @@ public class EditPostActivity extends AppCompatActivity implements
     private static final String STATE_KEY_EDITOR_FRAGMENT = "editorFragment";
     private static final String STATE_KEY_DROPPED_MEDIA_URIS = "stateKeyDroppedMediaUri";
     private static final String STATE_KEY_POST_LOCAL_ID = "stateKeyPostModelLocalId";
+    private static final String STATE_KEY_POST_REMOTE_ID = "stateKeyPostModelRemoteId";
     private static final String STATE_KEY_IS_NEW_POST = "stateKeyIsNewPost";
 
     private static int PAGE_CONTENT = 0;
@@ -319,14 +320,19 @@ public class EditPostActivity extends AppCompatActivity implements
                 mPost.setStatus(PostStatus.PUBLISHED.toString());
             } else if (extras != null) {
                 // Load post passed in extras
-                initializePostObjects(extras.getInt(EXTRA_POST_LOCAL_ID));
+                initializePostObjectsWithLocalId(extras.getInt(EXTRA_POST_LOCAL_ID));
             }
         } else {
             mDroppedMediaUris = savedInstanceState.getParcelable(STATE_KEY_DROPPED_MEDIA_URIS);
             mIsNewPost = savedInstanceState.getBoolean(STATE_KEY_IS_NEW_POST, false);
 
-            if (savedInstanceState.containsKey(STATE_KEY_POST_LOCAL_ID)) {
-                initializePostObjects(savedInstanceState.getInt(STATE_KEY_POST_LOCAL_ID));
+            // if we have a remote id saved, let's first try with that, as the local Id might have changed
+            // after FETCH_POSTS
+            if (savedInstanceState.containsKey(STATE_KEY_POST_REMOTE_ID)) {
+                initializePostObjectsWithRemoteId(savedInstanceState.getLong(STATE_KEY_POST_REMOTE_ID));
+            }
+            else if (savedInstanceState.containsKey(STATE_KEY_POST_LOCAL_ID)) {
+                initializePostObjectsWithLocalId(savedInstanceState.getInt(STATE_KEY_POST_LOCAL_ID));
             }
 
             mEditorFragment = (EditorFragmentAbstract) fragmentManager.getFragment(savedInstanceState, STATE_KEY_EDITOR_FRAGMENT);
@@ -404,16 +410,27 @@ public class EditPostActivity extends AppCompatActivity implements
         ActivityId.trackLastActivity(ActivityId.POST_EDITOR);
     }
 
-    private void initializePostObjects(int localPostId) {
+    private void initializePostObjectsWithLocalId(int localPostId) {
         mPost = mPostStore.getPostByLocalPostId(localPostId);
         if (mPost != null) {
-            mOriginalPost = mPost.clone();
-            mPost = UploadService.updatePostWithCurrentlyCompletedUploads(mPost);
-            mMediaMarkedUploadingOnStartIds =
-                    AztecEditorFragment.getMediaMarkedUploadingInPostContent(this, mPost.getContent());
-            Collections.sort(mMediaMarkedUploadingOnStartIds);
-            mIsPage = mPost.isPage();
+            initializePostObject();
         }
+    }
+
+    private void initializePostObjectsWithRemoteId(long remotePostId) {
+        mPost = mPostStore.getPostByRemotePostId(remotePostId, mSite);
+        if (mPost != null) {
+            initializePostObject();
+        }
+    }
+
+    private void initializePostObject() {
+        mOriginalPost = mPost.clone();
+        mPost = UploadService.updatePostWithCurrentlyCompletedUploads(mPost);
+        mMediaMarkedUploadingOnStartIds =
+                AztecEditorFragment.getMediaMarkedUploadingInPostContent(this, mPost.getContent());
+        Collections.sort(mMediaMarkedUploadingOnStartIds);
+        mIsPage = mPost.isPage();
     }
 
     // this method aims at recovering the current state of media items if they're inconsistent within the PostModel.
@@ -532,6 +549,7 @@ public class EditPostActivity extends AppCompatActivity implements
         // Saves both post objects so we can restore them in onCreate()
         savePostAsync(null);
         outState.putInt(STATE_KEY_POST_LOCAL_ID, mPost.getId());
+        outState.putLong(STATE_KEY_POST_REMOTE_ID, mPost.getRemotePostId());
         outState.putBoolean(STATE_KEY_IS_NEW_POST, mIsNewPost);
         outState.putSerializable(WordPress.SITE, mSite);
 

